src/libfsm/minimise.c

```diff
@@ -7,49 +7,660 @@
+				    ? li.labels[l_i]
+				    : env.dfa_labels[l_i]);
+
+				env.steps++;
+				if (try_partition(&env, label,
+					ec_src, ec_dst, pcounts)) {
+					int should_restart_EC;
+#if LOG_PARTITIONS > 0
+					fprintf(stderr, "# partition: ec_i %lu/%u/%u, l_i %lu/%u%s, pcounts [ %lu, %lu ]\n",
+					    ec_i, env.done_ec_offset, env.ec_count,
+					    l_i, li.limit, li.has_labels_for_EC ? "*" : "",
+					    pcounts[0], pcounts[1]);
+#endif
+
+					should_restart_EC =
+					    update_after_partition(&env,
+						pcounts, ec_src, ec_dst);
+
+					changed = 1;
+					env.ec_count++;
+					assert(env.done_ec_offset <= env.ec_count);
+					if (should_restart_EC) {
+						goto restart_EC;
+					}
+				}
+
+#if EXPENSIVE_INTEGRITY_CHECKS
+				check_done_ec_offset(&env);
+#endif
+			}
+		}
+
+		env.iter++;
+	} while (changed);
+
+	/* When all of the original input states are final, then the
+	 * initial not-final EC will be empty. Skip it in the mapping,
+	 * otherwise an unnecessary state will be created for it. */
+	ec_offset = (env.ecs[INIT_EC_NOT_FINAL] == NO_ID) ? 1 : 0;
+
+	/* Build map condensing to unique states */
+	for (i = ec_offset; i < env.ec_count; i++) {
+		fsm_state_t cur = MASK_EC_HEAD(env.ecs[i]);
+		while (cur != NO_ID) {
+			mapping[cur] = i - ec_offset;
+			cur = env.jump[cur];
+		}
+	}
+
+	if (minimized_state_count != NULL) {
+		*minimized_state_count = env.ec_count - ec_offset;
+	}
+
+#if LOG_MAPPINGS
+	for (i = 0; i < fsm->statecount; i++) {
+		fprintf(stderr, "# minimised_mapping[%lu]: %u\n",
+		    i, mapping[i]);
+	}
+#endif
+
+#if LOG_STEPS
+	fprintf(stderr, "# done in %lu iteration(s), %lu step(s), %ld -> %ld states, label_count %lu\n",
+            env.iter, env.steps, fsm->statecount,
+	    (size_t)(env.ec_count - ec_offset), env.dfa_label_count);
+#endif
+
+	res = 1;
+	/* fall through */
+
+cleanup:
+	f_free(fsm->opt->alloc, env.ecs);
+	f_free(fsm->opt->alloc, env.state_ecs);
+	f_free(fsm->opt->alloc, env.jump);
+	return res;
+}
+
+static void
+dump_ecs(FILE *f, const struct min_env *env)
+{
+#if LOG_ECS
+	size_t i = 0;
+	fprintf(f, "# iter %ld, steps %lu\n", env->iter, env->steps);
+	while (i < env->ec_count) {
+		fsm_state_t cur = MASK_EC_HEAD(env->ecs[i]);
+		fprintf(f, "M_EC[%lu]:", i);
+		while (cur != NO_ID) {
+			fprintf(f, " %u", cur);
+			cur = env->jump[cur];
+		}
+		fprintf(f, "\n");
+		i++;
+	}
+#else
+	(void)f;
+	(void)env;
+#endif
+}
+
+static void
+populate_initial_ecs(struct min_env *env, const struct fsm *fsm)
+{
+	size_t i;
+	for (i = 0; i < fsm->statecount; i++) {
+		const fsm_state_t ec = fsm_isend(fsm, i)
+		    ? INIT_EC_FINAL : INIT_EC_NOT_FINAL;
+		env->state_ecs[i] = ec;
+		/* link at head of the list */
+		env->jump[i] = env->ecs[ec];
+		env->ecs[ec] = i;
+#if LOG_INIT
+		fprintf(stderr, "# --init[%lu]: ec %d, jump[]= %d\n",
+		    i, ec, env->jump[i]);
+#endif
+	}
+
+	/* The dead state is not a member of any EC. */
+	env->state_ecs[env->dead_state] = NO_ID;
+}
+
+#if EXPENSIVE_INTEGRITY_CHECKS
```

Comment:
Great idea, thanks

Reply:
Do you think it's worth keeping this separate to `-DNDEBUG`? We have debugging assertions that are already super expensive, and I don't expect a `make DEBUG=1` build to perform well at all.

Reply:
This would be my vote if I may cast my possibly worthless two cents.

Reply:
Of course you may!! Always happy to have your input!

Reply:
On my own projects, I prefer to be more explicit and use `assert` for checks I expect to always run, and conditionally compile other more expensive per-loop invariants behind a `#define` like this. I very rarely use `-DNDEBUG`, because the performance gains in doing so have always been small (after wrapping asserts I know will be expensive), and it usually leads to unused variable warnings. Generally, I would wrap any assertion that would have a non-constant cost (such as sweeping through a collection to make sure something is present/removed) in an `#if` like this, but keep assertions on for things like bounds or NULL checks.

I left it this way so we could have this conversation during code review. The property tests aren't as meaningful with `-DNDEBUG`, though -- by design they're thrashing around in the state space and trying to provoke assert failures being used as internal consistency checks, which often catch problems quicker than tests on the results, so if we do those should be built separately.

Reply:
This is an excellent counterpoint. At the same time, I feel like EXPENSIVE_INTEGRITY_CHECKS might be not the best name for this (in that it is perhaps too specific). If NDEBUG is too generic, perhaps there's a suitable compromise? Or even something super obvious like -DPROPERTY_TESTS

Reply:
I don't think of them as being specific to the property tests, though. They're integrity, internal consistency, or invariant checks that can affect the algorithmic order of growth, rather than constant-time checks whose overhead can (hopefully) mostly be eliminated by branch prediction. Maybe `EXPENSIVE_ASSERTS`?

Reply:
My $.02, for what they're worth... 

afaik, libfsm doesn't have checks that would match the role of `EXPENSIVE_INTEGRITY_CHECKS`/`EXPENSIVE_ASSERTS` in any other place.  I'd rather keep the existing name, push for more expensive checks like these find their way into the code, and then try to generalize.  It's easier to find a better name when you have more than one concrete example.

Reply:
I had these two in mind when I wrote the above:
- https://github.com/katef/libfsm/blob/main/src/adt/edgeset.c#L237
- https://github.com/katef/libfsm/blob/main/src/adt/edgeset.c#L472

and there are others, such as:
- https://github.com/katef/libfsm/blob/main/src/libfsm/determinise.c#L88
- https://github.com/katef/libfsm/blob/main/src/libfsm/shortest.c#L113
- https://github.com/katef/libfsm/blob/main/src/libfsm/minimise.c#L27

Reply:
Hmm... you're right.  Many of those are quite expensive.

`EXPENSIVE_ASSERT()` would be a very clear way to write them, and would also let us have builds that can choose between costly assertion behavior, still having some assertion guard rails, and turning off assertions.